[DO NOT MERGE][LEARNER-316] WL Program About Page - move templates from theme repo

marjev opened the pull request:
### Reviewers:
- [ ] @edx/helio This is ready for the review.
- [x] @cptvitamin do you mind reviewing this from a11y standpoint?
- [x] @douglashall, perhaps you could review this from the WL standpoint?
- [ ] Product review

### Sandbox:
https://edx.sandbox.edx.org/programs/7f5526fc-07bd-45af-b65c-2f3a6d6e098c/about

### To-do:

- [x] Move the program marketing html templates from the themes repo
- [x] Mark strings for translation
- [x] Make sure the Course Catalog is used as the source of data (#14986)
- [x] Move related images from the themes repo
- [x] Move related sass/css files from the themes repo
- [x] Move related js files from the themes repo
- [x] Implement the design changes requested in https://openedx.atlassian.net/browse/LEARNER-312
- [x] write an acceptance test
- [x] provide screenshots (both desktop and mobile)
- [x] provide notes on how to implement and apply a theme

### Screenshots:
#### Desktop
![wl page - desktop](https://cloud.githubusercontent.com/assets/6833568/26450339/3ef5ae6e-4157-11e7-8707-3e05085ffda2.png)

#### Mobile
![mobile screenshot](https://cloud.githubusercontent.com/assets/6833568/26450386/6fb23400-4157-11e7-9f90-b1a5a3539c3c.png)

edx-webhook commented:
Thanks for the pull request, @marjev! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=14927"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

marjev commented:
jenkins run bokchoy

cptvitamin commented:
@marjev I would love to take a look!  Can you put this up on a sandbox somewhere?

marjev commented:
@cptvitamin Here's a link leading to the page hosted on a sandbox: https://edx.sandbox.edx.org/programs/7f5526fc-07bd-45af-b65c-2f3a6d6e098c/about

Also, I've shared with you a document containing the credentials necessary to access the sandbox.

cptvitamin commented:
@marjev I see that the grey on white text used in the cards on the bottom does not meet minimum contrast ratios.  

There is also no focus indicators on any of the focusable content in the main content region (the Buy button, the link to the bundle of courses, the controls to expand collapse the accordions, and the links in the cards.)  Remove outline:none from the CSS to fix that. Focus needs to be indicated on the accordion tabs when keyboard users use the arrow keys to navigate as well.  

I wasn't able to test the widget under Instructors in the Right hand sidebar because I'm not sure what its supposed to do.  It looks to me like it will require some specific accessibility accommodations. If you describe the expected use case and behaviors, I should be able to provide more guidance.

Also, there is no way to close the video modal with a keyboard.  You need to add a listener for the ESC key that closes it.  It also needs to have a close button.

Also, course cards looked funny when viewport is set to mobile sized (Chrome on Mac):
![screen shot showing extremely narrow course cards](https://cloud.githubusercontent.com/assets/444368/26203708/4ae3d6e0-3baa-11e7-87af-baf4c8cdc208.png)

Let me know when you get those issues fixed.  I'd like to take another look at this one.

marjev commented:
@cptvitamin, thanks for taking a look!

I'm going to fix the issues regarding the colors and the mobile design. I just noticed those.

I'll get back to you on the instructors widget. I don't have the concrete use case, since I just had to move the existing templates/js/css from the themes repo into the LMS repo.

douglashall commented on lms/djangoapps/commerce/utils.py:
```diff
@@ -60,10 +60,19 @@ def payment_page_url(self):
 
-    def checkout_page_url(self, sku):
+    def checkout_page_url(self, skus):
         """ Construct the URL to the ecommerce checkout page and include a product.
```
`""" Construct the URL to the ecommerce checkout page and include products.`

douglashall commented on lms/templates/courseware/_program_type_faq.html:
```diff
@@ -0,0 +1,50 @@
+        <div class="icon-container">
+          <img src="${static.url('images/img-2.png')}"/>
+        </div>
+        <small>${_("Learn Great Things at Your Own Place")}</small>
```
s/Place/Pace/

douglashall commented on lms/templates/courseware/_program_type_faq.html:
```diff
@@ -0,0 +1,50 @@
+  </div>
+  <div class="accordion-content">
+    <p>
+      ${_("Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt  ...
```
How are we going to populate this section of content with information about the program type?

marjev replied:
I've moved this from the themes repo. How's the FAQ data currently being populated?

douglashall replied:
Since we were originally implementing this page in individual themes and at least initially each site would only have a single program type associated with it, the plan was to hard code the real content here once the partner settle on a description of the program type. Since we are moving this to edx-platform, we will need to find a new solution for populating this section.

marjev replied:
@douglashall Can we remove this from the edx-platform and leave it as an optional accordion  item that will be defined in the themes repo?

That way, we would be able to solve this as part of a separate story, once we have an idea how we want to solve this problem.

douglashall replied:
Yes, that makes sense to me. Create an empty partial template that you include here. That way the theme just has to override that empty partial template.

douglashall commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -565,21 +555,3 @@ def _collect_course_pricing(self, course):
-            # Deduplicate program instructors using instructor name
-            self.instructors.update(
-                {instructor.get('name'): instructor for instructor in course_instructors.get('instructors', [])}
-            )
```
Why are we removing this? How are WL partners going to populate instructor data for courses?

marjev replied:
I'll have to tag @clintonb on this.
As I understood, data would be fetched from the Course Discovery. Program staff members would be used instead of instructors.

douglashall replied:
That makes sense to me, but if we are going to go that route, we need to have a plan for how WL partners will be able to author the instructor data and get it into the catalog.

marjev replied:
I will revert this change until we agree on the alternative. I don't want to break any existing use cases.

jzoldak commented:
jenkins run bokchoy
jenkins run lettuce
jenkins run python

marjev commented:
jenkins run all

marjev commented:
@douglashall I think we've addressed most of your comments. Care to take another look?

@cptvitamin I've addressed some of the issues you've shed the light on. Could you please take another look at https://edx.sandbox.edx.org/programs/7f5526fc-07bd-45af-b65c-2f3a6d6e098c/about.

You should be able to see the instructor data and get the feeling of how the instructors widget should work. If you have any questions, please let me know.

douglashall commented on lms/djangoapps/commerce/utils.py:
```diff
@@ -76,10 +76,19 @@ def payment_page_url(self):
 
-    def checkout_page_url(self, sku):
-        """ Construct the URL to the ecommerce checkout page and include a product.
+    def checkout_page_url(self, skus):
```
How about using `*args` here instead of requiring a list be passed?

douglashall commented on lms/djangoapps/commerce/utils.py:
```diff
@@ -76,10 +76,19 @@ def payment_page_url(self):
-        return "{}?sku={}".format(self.get_absolute_ecommerce_url(self.config.single_course_checkout_page), sku)
+        return '{checkout_page_path}?{skus}'.format(
+            checkout_page_path=self.get_absolute_ecommerce_url(self.config.single_course_checkout_page),
+            skus='&'.join('sku={sku}'.format(sku=sku) for sku in skus),
```
Use [urlencode](https://docs.python.org/2/library/urllib.html#urllib.urlencode).

douglashall commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -508,9 +508,15 @@ def _extend_program(self):
             uuid=self.data['uuid']
         )
         program_instructors = cache.get(cache_key)
+        ecommerce_service = EcommerceService()
```
Move this to `__init__`.

douglashall commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -508,9 +508,15 @@ def _extend_program(self):
+        applicable_seat_types = self.data['applicable_seat_types']
 
         for course in self.data['courses']:
+            for seat in course['course_runs'][0]['seats']:
```
I know we are only supporting single-click program purchase for programs that consist of courses with a single course run, but do we actually need to restrict that here?

marjev replied:
@vkaracic please remind me, we're fetching the 1st course run because it's the latest course run of the course, right?

vkaracic replied:
Actually I'm not sure if the runs are sorted. This might be an issue.

marjev commented:
@edx/helio please review.

marjev commented:
jenkins run bokchoy

vkaracic commented on lms/djangoapps/commerce/tests/test_utils.py:
```diff
@@ -91,8 +91,12 @@ def test_get_receipt_page_url(self):
-    def test_checkout_page_url(self):
+    @ddt.data(
+        (['TESTSKU'], 'http://ecommerce_url/test_basket/?sku=TESTSKU'),
+        (['TESTSKU1', 'TESTSKU2', 'TESTSKU3'], 'http://ecommerce_url/test_basket/?sku=TESTSKU1&sku=TESTSKU2&sku=TESTSKU3')
```
Would it be better to have the `http://ecommerce_url` as a constant here?

clintonb replied:
Build the URL from the SKUs and the setting `ECOMMERCE_PUBLIC_URL_ROOT`.

vkaracic commented on lms/djangoapps/commerce/utils.py:
```diff
@@ -76,10 +77,16 @@ def payment_page_url(self):
 
-    def checkout_page_url(self, sku):
-        """ Construct the URL to the ecommerce checkout page and include a product.
+    def checkout_page_url(self, *args):
```
I would rename `args` to `skus` here seeing that only SKUs should be sent to this method. Also add it to the docstring.

vkaracic commented on lms/static/js/leanModal.js:
```diff
@@ -0,0 +1,143 @@
+         * will add a .click() handler that properly opens the dialog.
+         *
+         * The launching element must:
+         *   - be an <a> element, not a button,
```
I think this part should be opposite. It should be a button not an anchor.

vkaracic commented on lms/static/js/leanModal.js:
```diff
@@ -0,0 +1,143 @@
+                    }
+
+                    $(document).on('keydown.leanModal', function(event) {
+                        if (event.which === 27) {
```
Add a comment explaining that this checks if the escape button is pressed.

vkaracic commented on lms/static/js/leanModal.js:
```diff
@@ -0,0 +1,143 @@
+                            opacity: 0,
+                            'z-index': 11000,
+                            left: 50 + '%',
+                            'margin-left': -($(modalId).outerWidth() / 2) + 'px',
```
Should we be consistent with the quotes on all the keys? You're call.

vkaracic commented on lms/static/js/leanModal.js:
```diff
@@ -0,0 +1,143 @@
+
+    $(document).ready(function($) {  // eslint-disable-line no-shadow
+        $('button[rel*=leanModal]').each(function() {
+            var sep, embed;
```
Please change s/sep/separator to descriptive.

vkaracic commented on lms/static/js/program_marketing.js:
```diff
@@ -0,0 +1,167 @@
+    $('.course-card').toggleClass('slidable', isMobileResolution);
+    if (isMobileResolution) {    // Second condition will avoid the multiple calls from resize
+        $('.copy-meta-mobile').show();
+        $('.copy-meta').hide();
```
Couldn't have this been done with CSS?

clintonb replied:
Yes. Use CSS, preferably Bootstrap.

vkaracic commented on lms/templates/courseware/program_marketing.html:
```diff
@@ -1,3 +1,404 @@
+        <div class="col col-12 md-col-8 lg-col-9">
+          <div class="org-logo">
+            % if authoring_organizations and authoring_organizations[0]['logo_image_url']:
+              <img src="${authoring_organizations[0]['logo_image_url']}" alt="${authoring_organizations[0]['name']}">
```
Since `authoring_organizations` is list, should this be in a for-loop?

clintonb replied:
Yes.

vkaracic commented on lms/templates/courseware/program_marketing.html:
```diff
@@ -1,3 +1,404 @@
+          </a>
+        </div>
+        <div class="col col-12 md-col-4 lg-col-3" id="course-trailer">
+          <a href="#video-modal" class="media trailer-link visible-sm" rel="leanModal">
```
Should this be a `button` element, a11y-wise? Since it is opening a modal.

vkaracic commented on lms/templates/courseware/program_marketing.html:
```diff
@@ -1,3 +1,404 @@
+          </a>
+          <div class="hidden-sm">
+            <button href="#video-modal" class="media btn-play" rel="leanModal">
+              <span class="sr">Play the ${title} program video</span>
```
This should also be translated (coming from the that added it).

vkaracic commented on lms/templates/courseware/program_marketing.html:
```diff
@@ -1,3 +1,404 @@
+              <ul class="list-unstyled faq-links-list">
+                % for item in faq:
+                  <li class="item">
+                    <a href="#!">${item['question']}</a>
```
This should probably be a button as well, but I don't know how easy of a change it could be.

vkaracic commented on openedx/core/djangoapps/catalog/tests/factories.py:
```diff
@@ -112,19 +121,56 @@ class CourseFactory(DictFactoryBase):
+
+
+class PersonFactory(DictFactoryBase):
+    bio = factory.fuzzy.FuzzyText()
```
You could probably use [ `Faker('paragraphs')` ](https://faker.readthedocs.io/en/master/providers/faker.providers.lorem.html)

clintonb commented on common/test/acceptance/pages/lms/programs.py:
```diff
@@ -31,3 +31,32 @@ class ProgramDetailsPage(PageObject):
+    @property
+    def are_instructor_profiles_present(self):
+        """Check whether instructor profiles are present."""
+        return
```
This needs a comment explaining why it does nothing.

clintonb commented on common/test/acceptance/pages/lms/programs.py:
```diff
@@ -31,3 +31,32 @@ class ProgramDetailsPage(PageObject):
+        return self.q(css='#program-details-page').present
+
+    @property
+    def are_course_cards_present(self):
```
These properties are only used in one place. Do they really need to be properties, or can they just be calls in the consuming test?

clintonb commented on common/test/acceptance/tests/lms/test_programs.py:
```diff
@@ -102,6 +102,33 @@ def test_enrollments_and_programs(self):
+
+        self.marketing_page.visit()
+
+        self.assertTrue(self.marketing_page.are_course_cards_present)
```
We should already have a unit test that can tell if you if the page will load. What's the value of checking if a few elements are on the page? Frankly, I'm not sure of the value of the test given that it doesn't check for actual program details.

clintonb commented on lms/djangoapps/commerce/tests/test_utils.py:
```diff
@@ -91,8 +91,12 @@ def test_get_receipt_page_url(self):
 
     @override_settings(ECOMMERCE_PUBLIC_URL_ROOT='http://ecommerce_url')
-    def test_checkout_page_url(self):
+    @ddt.data(
```
Why are these changes necessary for the program about page? This seems like it belongs in a separate commit.

clintonb commented on lms/static/js/leanModal.js:
```diff
@@ -0,0 +1,136 @@
+(function($) {  // eslint-disable-line wrap-iife
```
Please use Bootstrap.

marjev replied:
This is not as straightforward as I expected it to be. 
Bootstrap CSS is not included by default into main Django template and I can't find a way to reference it from within the project. I can reference externally hosted version but it breaks the existing styling.
We can file a tech debt ticket since this is nice-to-have.

Let's move with making it work and any nice-to-have's can be a separate task.

clintonb commented on lms/static/sass/_build-learner-dashboard.scss:
```diff
@@ -9,3 +9,4 @@
 @import 'elements/progress-circle';
 @import 'views/program-details';
 @import 'views/program-list';
+@import "views/program-marketing-page";
```
Why is this part of learner dashboard?

marjev replied:
Where should I put it?

clintonb replied:
A different file, perhaps its own. Where does the course detail page SCSS live?

marjev replied:
Source files are within `sass/course`.
Build file is called `_build-course.scss`.

clintonb commented on lms/templates/courseware/program_marketing.html:
```diff
@@ -1,3 +1,407 @@
+<%namespace name='static' file='../static_content.html'/>
+
+<%
+faq = program['faq']
```
Do we need variables for every one of these? Why not just access the entries in the dict?

clintonb commented on lms/templates/courseware/program_marketing.html:
```diff
@@ -1,3 +1,407 @@
 <%page expression_filter="h"/>
+<%inherit file="../main.html"/>
```
All new templates should be Django templates: http://open-edx-proposals.readthedocs.io/en/latest/oep-0011.html#technology-selection. Death to Mako!

clintonb commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -577,3 +582,37 @@ def _collect_instructors(self, course_run):
                 {instructor.get('name'): instructor for instructor in course_instructors.get('instructors', [])}
             )
+
+    def _collect_one_click_purchase_eligibility_data(self):
```
Add comments explaining the purpose of this method.

clintonb commented on openedx/core/djangoapps/programs/utils.py:
```diff
@@ -577,3 +582,37 @@ def _collect_instructors(self, course_run):
+                log.exception('Failed to get discount price for following product SKUs: %s ', ', '.join(skus))
+
+        self.data.update({
+            'basket_page_url': self.ecommerce_service.checkout_page_url(*skus) if skus else '#courses',
```
Code smell! Why is there an if statement here? Your data model definitely should *not* care about an element ID from the view.

marjev commented:
I'll split this PR into smaller ones.
